src/geom/elem.C

```diff
@@ -3404,4 +3404,37 @@ Elem::is_internal(const unsigned int i) const
   }
 }
 
+bool
+Elem::edge_orientation(const unsigned int i) const
+{
+  libmesh_assert_less (i, this->n_edges());
+
+  return this->point(this->local_edge_node(i, 0)) >
+         this->point(this->local_edge_node(i, 1));
+}
+
+bool
+Elem::face_orientation(const unsigned int i) const
+{
+  libmesh_assert_less (i, this->n_faces());
```

Comment:
I think we should assert that the Elem's dimension == 3 here, and I'm actually in favor of making it an opt-mode `libmesh_error_msg_if()` check instead of just an assert. If this function is called on a 2D element (in opt mode), it's possible for N == 2, at which point computing N - 3 will wrap to a huge number in the loops below, likely leading to a segfault.

Reply:
> I think we should assert that the Elem's dimension == 3 here

That's implicit since `n_faces` is zero if that's not the case.

> I'm actually in favor of making it an opt-mode `libmesh_error_msg_if()` check instead of just an assert.

Help me here, how do you decide? I can similarly segfault on a call to `local_edge_node` if I ask for a non-existent edge.

Reply:
It's true that we should be consistent (if we assert in other places, we should assert here too) and performant (so, perfer debug-mode asserts to opt-mode checks) so I'm OK keeping it as an assert. 

I thought it would clearer to someone seeing an assert from this function if it said "You can only call Elem::positive_face_orientation() on 3D elements." instead of something about `i` not being less than the number of faces, but I don't feel to strongly about it.